apps/api/src/modules/notifications/queues/queue.service.ts

```diff
@@ -0,0 +1,179 @@
+
+    return queue;
+  }
+
+  getOrCreateQueue(action: string, providerType: string, providerId: string): Queue {
+    const queueName = `${action}-${providerType}-${providerId}`;
+
+    if (!this.queues.has(queueName)) {
+      this.logger.log(`Creating new queue and worker for ${queueName}`);
+      this.createQueue(queueName);
+      this.createWorker(action, providerType, providerId, queueName);
+    }
+
+    return this.queues.get(queueName);
+  }
+
+  private createWorker(
+    action: string,
+    providerType: string,
+    providerId: string,
+    queueName: string,
+  ): void {
+    const processJob = async (job): Promise<void> => {
+      switch (`${action}-${providerType}`) {
+        case `send-${ChannelType.SMTP}`:
+          await this.smtpNotificationConsumer.processSmtpNotificationQueue(job.data.id);
+          break;
+        case `send-${ChannelType.MAILGUN}`:
+          await this.mailgunNotificationConsumer.processMailgunNotificationQueue(job.data.id);
+          break;
+        case `send-${ChannelType.WA_360_DAILOG}`:
+          await this.wa360dialogNotificationConsumer.processWa360dialogNotificationQueue(
+            job.data.id,
+          );
+          break;
+        case `send-${ChannelType.WA_TWILIO}`:
+          await this.waTwilioNotificationConsumer.processWaTwilioNotificationQueue(job.data.id);
+          break;
+        case `send-${ChannelType.SMS_TWILIO}`:
+          await this.smsTwilioNotificationConsumer.processSmsTwilioNotificationQueue(job.data.id);
+          break;
+        case `send-${ChannelType.SMS_PLIVO}`:
+          await this.smsPlivoNotificationConsumer.processSmsPlivoNotificationQueue(job.data.id);
+          break;
+        case `send-${ChannelType.WA_TWILIO_BUSINESS}`:
+          await this.waTwilioBusinessNotificationConsumer.processWaTwilioBusinessNotificationQueue(
+            job.data.id,
+          );
+          break;
+        case `send-${ChannelType.SMS_KAPSYSTEM}`:
+          await this.smsKapsystemNotificationConsumer.processSmsKapsystemNotificationQueue(
+            job.data.id,
+          );
+          break;
+        default:
+          this.logger.error(
+            `Unsupported action-providerType combination: ${action}-${providerType}`,
+          );
+          return;
+      }
+    };
+
+    const worker = new Worker(queueName, processJob, { connection: this.redisConfig });
+    worker.on('completed', (job) => {
+      this.logger.log(JSON.stringify(job));
+      this.logger.log(`Job completed: ${job.id}`);
+    });
+    worker.on('failed', (job, err) => {
+      this.logger.log(JSON.stringify(job));
+      this.logger.error(`Job failed: ${job.id}`, err);
+    });
+
+    this.workers.set(queueName, worker);
+  }
+
+  private startCleanupTask(): void {
+    setInterval(() => {
+      this.cleanupIdleResources();
+    }, this.cleanupInterval);
+  }
+
+  private async cleanupIdleResources(): Promise<void> {
+    const now = Date.now();
```

Comment:
@LakshayaT  Did you test this? How was the behaviour?

Reply:
Tested it now and this is the response (increased cron time to every 10 seconds and cleanup time to every minute to reduce logs):

```bash
[OsmoX] 12573  - 7/2/2024, 10:26:24 AM     LOG [NestApplication] Nest application successfully started +5ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +6s
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [NotificationsService] Adding 3 pending notifications to queue +15ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [QueueService] Creating new queue and worker for send-2-3 +2ms
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [NotificationsService] Getting notification with id: 60 +20ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [QueueService] Creating new queue and worker for send-1-1 +4ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [MailgunNotificationConsumer] Sending notification with id: 60 +2ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [NotificationsService] Getting notification with id: 60 +0ms
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [NotificationsService] Getting notification with id: 62 +32ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [SmtpNotificationConsumer] Sending notification with id: 62 +2ms
[OsmoX] 12573  - 7/2/2024, 10:26:30 AM     LOG [NotificationsService] Getting notification with id: 62 +0ms
[OsmoX] 12573  - 7/2/2024, 10:26:31 AM     LOG [QueueService] {"name":"60","data":{"id":60,"providerId":3},"opts":{"attempts":0,"delay":0},"id":"1","progress":0,"stacktrace":[],"attemptsStarted":1,"attemptsMade":1,"delay":0,"timestamp":1719896190038,"queueQualifiedName":"bull:send-2-3","processedOn":1719896190043,"token":"8e0d534d-1a28-4abb-94fd-e8ef94292bb4:0","finishedOn":1719896191118} +1s
[OsmoX] 12573  - 7/2/2024, 10:26:31 AM     LOG [QueueService] Job completed: 1 +1ms
[OsmoX] 12573  - 7/2/2024, 10:26:33 AM     LOG [QueueService] {"name":"62","data":{"id":62,"providerId":1},"opts":{"attempts":0,"delay":0},"id":"8","progress":0,"stacktrace":[],"attemptsStarted":1,"attemptsMade":1,"delay":0,"timestamp":1719896190057,"queueQualifiedName":"bull:send-1-1","processedOn":1719896190084,"token":"cbf1ea92-9516-47a2-97a2-caeb864887f4:0","finishedOn":1719896193488} +2s
[OsmoX] 12573  - 7/2/2024, 10:26:33 AM     LOG [QueueService] Job completed: 8 +0ms
[OsmoX] 12573  - 7/2/2024, 10:26:33 AM     LOG [NotificationsService] Getting notification with id: 63 +1ms
[OsmoX] 12573  - 7/2/2024, 10:26:33 AM     LOG [SmtpNotificationConsumer] Sending notification with id: 63 +1ms
[OsmoX] 12573  - 7/2/2024, 10:26:33 AM     LOG [NotificationsService] Getting notification with id: 63 +0ms
[OsmoX] 12573  - 7/2/2024, 10:26:35 AM     LOG [QueueService] {"name":"63","data":{"id":63,"providerId":1},"opts":{"attempts":0,"delay":0},"id":"9","progress":0,"stacktrace":[],"attemptsStarted":1,"attemptsMade":1,"delay":0,"timestamp":1719896190094,"queueQualifiedName":"bull:send-1-1","processedOn":1719896193488,"token":"cbf1ea92-9516-47a2-97a2-caeb864887f4:0","finishedOn":1719896195651} +2s
[OsmoX] 12573  - 7/2/2024, 10:26:35 AM     LOG [QueueService] Job completed: 9 +0ms
[OsmoX] 12573  - 7/2/2024, 10:26:40 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +4s
[OsmoX] 12573  - 7/2/2024, 10:26:40 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 12573  - 7/2/2024, 10:26:40 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +2ms
[OsmoX] 12573  - 7/2/2024, 10:26:50 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 12573  - 7/2/2024, 10:26:50 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 12573  - 7/2/2024, 10:26:50 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +3ms
[OsmoX] 12573  - 7/2/2024, 10:27:00 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 12573  - 7/2/2024, 10:27:00 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 12573  - 7/2/2024, 10:27:00 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +2ms
[OsmoX] 12573  - 7/2/2024, 10:27:10 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 12573  - 7/2/2024, 10:27:10 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 12573  - 7/2/2024, 10:27:10 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +2ms
[OsmoX] 12573  - 7/2/2024, 10:27:20 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 12573  - 7/2/2024, 10:27:20 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 12573  - 7/2/2024, 10:27:20 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +3ms
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Removing idle queue and worker: send-2-3 +4s
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Worker removed for queue: send-2-3 due to inactivity. +22ms
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Removing idle queue and worker: send-1-1 +1ms
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Queue removed: send-2-3 due to inactivity. +1ms
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Queue events removed: send-2-3 due to inactivity. +1ms
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Worker removed for queue: send-1-1 due to inactivity. +1ms
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Queue removed: send-1-1 due to inactivity. +0ms
[OsmoX] 12573  - 7/2/2024, 10:27:23 AM     LOG [QueueService] Queue events removed: send-1-1 due to inactivity. +1ms
[OsmoX] 12573  - 7/2/2024, 10:27:30 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +6s
[OsmoX] 12573  - 7/2/2024, 10:27:30 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 12573  - 7/2/2024, 10:27:30 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +3ms
```

Reply:
@LakshayaT  You need to add notifications and check if it is creating them again properly when in use after getting removed.
We probably don't need to enable this part right no.
For now create a Env Variable which will decide whether we want to do cleanup or not, also keep the timeouts too  in env with some default values ideally we will probably keep them like 1 week or 1 month like that, if its inactive for such a long time we remove them.

Reply:
> You need to add notifications and check if it is creating them again properly when in use after getting removed.

Yes, the queue is recreated when a new notification is added/is pending:

```bash
[OsmoX] 14232  - 7/2/2024, 10:37:50 AM     LOG [NestApplication] Nest application successfully started +5ms
[OsmoX] 14232  - 7/2/2024, 10:38:00 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +9s
[OsmoX] 14232  - 7/2/2024, 10:38:00 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 14232  - 7/2/2024, 10:38:00 AM     LOG [NotificationsService] Adding 1 pending notifications to queue +7ms
[OsmoX] 14232  - 7/2/2024, 10:38:00 AM     LOG [QueueService] Creating new queue and worker for send-1-1 +4ms
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
[OsmoX] 14232  - 7/2/2024, 10:38:00 AM     LOG [NotificationsService] Getting notification with id: 62 +28ms
[OsmoX] 14232  - 7/2/2024, 10:38:00 AM     LOG [SmtpNotificationConsumer] Sending notification with id: 62 +6ms
[OsmoX] 14232  - 7/2/2024, 10:38:00 AM     LOG [NotificationsService] Getting notification with id: 62 +1ms
[OsmoX] 14232  - 7/2/2024, 10:38:02 AM     LOG [QueueService] {"name":"62","data":{"id":62,"providerId":1},"opts":{"attempts":0,"delay":0},"id":"10","progress":0,"stacktrace":[],"attemptsStarted":1,"attemptsMade":1,"delay":0,"timestamp":1719896880032,"queueQualifiedName":"bull:send-1-1","processedOn":1719896880040,"token":"ed8a0b14-813f-46d2-9888-e5c53e990774:0","finishedOn":1719896882424} +2s
[OsmoX] 14232  - 7/2/2024, 10:38:02 AM     LOG [QueueService] Job completed: 10 +1ms
[OsmoX] 14232  - 7/2/2024, 10:38:10 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +8s
[OsmoX] 14232  - 7/2/2024, 10:38:10 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 14232  - 7/2/2024, 10:38:10 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +1ms
[OsmoX] 14232  - 7/2/2024, 10:38:20 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 14232  - 7/2/2024, 10:38:20 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 14232  - 7/2/2024, 10:38:20 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +4ms
[OsmoX] 14232  - 7/2/2024, 10:38:30 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 14232  - 7/2/2024, 10:38:30 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 14232  - 7/2/2024, 10:38:30 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +4ms
[OsmoX] 14232  - 7/2/2024, 10:38:40 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 14232  - 7/2/2024, 10:38:40 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 14232  - 7/2/2024, 10:38:40 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +3ms
[OsmoX] 14232  - 7/2/2024, 10:38:50 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 14232  - 7/2/2024, 10:38:50 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 14232  - 7/2/2024, 10:38:50 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +4ms
[OsmoX] 14232  - 7/2/2024, 10:38:50 AM     LOG [QueueService] Removing idle queue and worker: send-1-1 +845ms
[OsmoX] 14232  - 7/2/2024, 10:38:50 AM     LOG [QueueService] Worker removed for queue: send-1-1 due to inactivity. +1ms
[OsmoX] 14232  - 7/2/2024, 10:38:50 AM     LOG [QueueService] Queue removed: send-1-1 due to inactivity. +1ms
[OsmoX] 14232  - 7/2/2024, 10:38:50 AM     LOG [QueueService] Queue events removed: send-1-1 due to inactivity. +0ms
[OsmoX] 14232  - 7/2/2024, 10:39:00 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +9s
[OsmoX] 14232  - 7/2/2024, 10:39:00 AM     LOG [NotificationsService] Getting all active pending notifications +0ms
[OsmoX] 14232  - 7/2/2024, 10:39:00 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +2ms
[OsmoX] 14232  - 7/2/2024, 10:39:10 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +10s
[OsmoX] 14232  - 7/2/2024, 10:39:10 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 14232  - 7/2/2024, 10:39:10 AM     LOG [NotificationsService] Adding 1 pending notifications to queue +4ms
[OsmoX] 14232  - 7/2/2024, 10:39:10 AM     LOG [QueueService] Creating new queue and worker for send-1-1 +3ms
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
It is highly recommended to use a minimum Redis version of 6.2.0
             Current: 6.0.16
[OsmoX] 14232  - 7/2/2024, 10:39:10 AM     LOG [NotificationsService] Getting notification with id: 62 +11ms
[OsmoX] 14232  - 7/2/2024, 10:39:10 AM     LOG [SmtpNotificationConsumer] Sending notification with id: 62 +1ms
[OsmoX] 14232  - 7/2/2024, 10:39:10 AM     LOG [NotificationsService] Getting notification with id: 62 +1ms
[OsmoX] 14232  - 7/2/2024, 10:39:12 AM     LOG [QueueService] {"name":"62","data":{"id":62,"providerId":1},"opts":{"attempts":0,"delay":0},"id":"11","progress":0,"stacktrace":[],"attemptsStarted":1,"attemptsMade":1,"delay":0,"timestamp":1719896950028,"queueQualifiedName":"bull:send-1-1","processedOn":1719896950030,"token":"a0f54640-fb4d-4a00-a3fe-d7cec1e0b40f:0","finishedOn":1719896952409} +2s
[OsmoX] 14232  - 7/2/2024, 10:39:12 AM     LOG [QueueService] Job completed: 11 +0ms
[OsmoX] 14232  - 7/2/2024, 10:39:20 AM     LOG [NotificationsService] Starting CRON job to add pending notifications to queue +8s
[OsmoX] 14232  - 7/2/2024, 10:39:20 AM     LOG [NotificationsService] Getting all active pending notifications +1ms
[OsmoX] 14232  - 7/2/2024, 10:39:20 AM     LOG [NotificationsService] Adding 0 pending notifications to queue +3ms
```

Setting up configuration for cleanup via env has been implemented in https://github.com/OsmosysSoftware/osmo-x/pull/264/commits/66057ff088f70d7c6c988b83d0e9e22881297b2c